src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -106,71 +109,110 @@ private static void instrumentMatch(ConcurrentHashMap<MatchInfo, AtomicInteger>
+            } else if (codeRange == CodeRange.CR_7BIT && regexp.encoding == Encodings.US_ASCII) {
+                asciiOnlyProfile.enter();
 
-        @Specialization(guards = {
-                "!isSameEncoding(regexp, encoding)",
-                "!isUSASCII(regexp, rope)",
-                "!isFixedEncoding(regexp, encoding)"
-        })
-        protected RubyEncoding fallback(RubyRegexp regexp, Rope rope, RubyEncoding encoding) {
-            return encoding;
-        }
+                return Encodings.US_ASCII;
+            } else if (!matchRope.getEncoding().isAsciiCompatible()) {
+                asciiIncompatibleMatchStringEncodingProfile.enter();
 
-        protected boolean isSameEncoding(RubyRegexp regexp, RubyEncoding encoding) {
-            return regexp.encoding == encoding;
-        }
+                return raiseEncodingCompatibilityError(regexp, matchRope);
+            } else if (regexp.options.isFixed()) {
+                fixedRegexpEncodingProfile.enter();
+
+                if (!regexp.encoding.jcoding.isAsciiCompatible() || codeRange != CodeRange.CR_7BIT) {
+                    asciiIncompatibleFixedRegexpEncodingProfile.enter();
+
+                    return raiseEncodingCompatibilityError(regexp, matchRope);
+                }
+
+                return regexp.encoding;
+            } else {
+                returnMatchStringEncodingProfile.enter();
+
+                if (regexp.options.isEncodingNone() && matchStringEncoding != Encodings.BINARY &&
+                        codeRange != CodeRange.CR_7BIT) {
+                    binaryRegexpMatchProfile.enter();
 
-        protected boolean isUSASCII(RubyRegexp regexp, Rope rope) {
-            return regexp.encoding == Encodings.US_ASCII &&
-                    codeRangeNode.execute(rope) == CodeRange.CR_7BIT;
+                    warnHistoricalBinaryRegexpMatch(matchStringEncoding);
+                }
+
+                return matchStringEncoding;
+            }
         }
 
-        protected boolean isFixedEncoding(RubyRegexp regexp, RubyEncoding encoding) {
-            return regexp.options.isFixed() && encoding.jcoding.isAsciiCompatible();
+        // MRI: reg_enc_error
+        private RubyEncoding raiseEncodingCompatibilityError(RubyRegexp regexp, Rope matchRope) {
+            throw new RaiseException(
+                    getContext(),
+                    coreExceptions().encodingCompatibilityErrorRegexpIncompatible(
+                            regexp.encoding.jcoding,
+                            matchRope.getEncoding(),
+                            this));
         }
 
+        private void warnHistoricalBinaryRegexpMatch(RubyEncoding matchStringEncoding) {
+            if (warnNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                warnNode = insert(new WarnNode());
+            }
+
+            // MRI warns unconditionally in this case. I.e., the warning flag does not need to be supplied.
```

Comment:
Okay. Thanks for flagging that. The problem I ran into when using `shouldWarn` is that the following would not print a warning:

```bash
$ jt ruby -e '/ae/n =~ "aeé"'
```

I needed to supply the `-w` flag to print the warning, which as your example shows, MRI does not require. Do we just have a different default warn level?

Reply:
Maybe you accidentally used `shouldWarnForDeprecation()` instead of `shouldWarn()`, or a WarningNode?
`WarnNode#shouldWarn()` should warn by default, since the default is `$VERBOSE = false`.

Reply:
It turns out the block of code above this that I added to optimize common cases needs to take the `NOENCODING` option into account. Maybe I did use `shouldWarnForDeprecation` originally -- I can't really remember -- but checking just now the problem is that the `warningMessage` call is not being encountered with the sample input you provided above.

Reply:
I thought NOENCODING //n implied FIXED/fixed_encoding? but apparently not:
```
irb(main):001:0> //n.fixed_encoding?
=> false
irb(main):002:0> //u.fixed_encoding?
=> true
```
Both can be checked at once with a bitmask though, and with some helper method on RegexpOptions like `#adaptEncoding()` or so.